docs/shared/recipes/running-tasks/customizing-inputs.md

```diff
@@ -0,0 +1,354 @@
+        "inputs": ["..."],
+        ...
+      }
+      ...
+    }
+  }
+}
+```
+
+{% /tab %}
+
+## Include all project files and dependencies in cache hash
+
+The following definition includes all files of the project itself as well as of all its dependencies in the cache hash, hence telling Nx to invalidate the cache whenever
+
+- any file of the project itself changes
+- any file of any dependency changes
+
+```jsonc {% fileName="nx.json" %}
+{
+  "namedInputs": {
+    "default": ["{projectRoot}/**/*"]
+  },
+  "targetDefaults": {
+    "build": {
+      "inputs": ["default", "^default"]
+    }
+  }
+}
+```
+
+This definition is the default behavior of Nx, even if you don't specify any `inputs` at all. This is a rather cautious approach. It will always give you correct output, but it might re-run a task in some cases where the cache could have been used instead.
+
+Note, Nx uses the [minimatch](https://github.com/isaacs/minimatch) library to process glob patterns.
+
+{% callout title="Replacements" %}
+
+Note how you can use `{projectRoot}` and `{workspaceRoot}` as placeholders to simplify writing glob definitions.
+
+{% /callout %}
+
+If you're wondering what `namedInputs` are, read the next section.
+
+## Reusing inputs definitions with namedInputs
+
+If you find yourself reusing the same `inputs` definitions, you can instead create a `namedInput`. It is like a variable definition which can then be reused in the `inputs` array.
+
+```jsonc {% fileName="nx.json" highlightLines=[3, 7] %}
+{
+  "namedInputs": {
+    "default": ["{projectRoot}/**/*"]
+  },
+  "targetDefaults": {
+    "build": {
+      "inputs": ["default", "^default"]
+    }
+  }
+}
+```
+
+The `^` character at the beginning of the `^default` string means this entry applies to the project dependencies of the project, not the project itself. In our example of `myreactapp` depending on `shared-ui`, if we run
+
+```shell
+nx build myreactapp
+```
+
+...then by having defined
+
+- `: ["default",...]` - it will invalidate the cache of `myreactapp` whenever some src file of `myreactapp` itself changes
+- `inputs: [..., "^default"]` - it will in addition invalidate the cache of `myreactapp` whenever a file of `shared-ui` (or any of its dependencies) changes
+
+## Exclude files from invalidating cache
+
+Sometimes you might want to exclude specific project files s.t. they don't invalidate the cache for a given target. For example, we want spec/test files to invalidate the `test` target (by explicitly including `*.spec.ts` files), but we might want to optimize and exclude them from our `build` target. Hence, whenever we just change a test file, our `build` cache would still be working.
+
+Here's how we could define that, starting from our default situation:
+
+```jsonc {% fileName="nx.json" highlightLines=[4, 5, 6, 7, 11] %}
+{
+  "namedInputs": {
+    "default": ["{projectRoot}/**/*"],
+    "production": [
+      "default",
+      "!{projectRoot}/**/?(*.)+(spec|test).[jt]s?(x)?(.snap)"
```

Comment:
```suggestion
      "!{projectRoot}/**/?(*.)+spec.ts"
```

Reply:
Simplified the regex so people aren't distracted by all the regex syntax when we want them to focus on the `!` at the front